### ISSUE TITLE
With only returns object with changed value - missing other properties

### ISSUE THREAD

**Issue by dnaustin** [ProductiveRage/Bridge.Immutable#11]
I am working through your Bridge.React tutorial on your blog and I have got to part 3. I have implemented all of the code up to pure components and have tested the app but it is failing the ValidateMessage check in the MessageWriterStore. From investigation, it appears that when the onChange of the MessageEditor is called, the props.Message.With returns a MessageEditState object that only has the title or content initialised depending on which textbox has been typed into.

``onChange: newTitle => this.props.OnChange(this.props.Message.With(_ => _.Title, new TextEditState(newTitle)))``

**Comment by ProductiveRage** [comment]
I have very recently upgraded this package to support Bridge 17 and I was worried that a new error had crept in.. but I've tried to construct a minimum reproduce case to match what you're describing and I can't.

To do so, I created a completely fresh solution with a Class Library project and added the ProductiveRage.Immutable NuGet package (which also brings in the Bridge package) and changed the Class1.cs file to the following:

```
using System;
using ProductiveRage.Immutable;

namespace BridgeImmutable17Test
{
    public class Class1
    {
        static void Main()
        {
            Console.WriteLine("Hi!");
            var message11 = new MessageDetails("Title1", "Content1");
            var message21 = message11.With(_ => _.Title, "Title2");
            var message22 = message21.With(_ => _.Content, "Content2");
            Console.WriteLine("message11: " + message11.Title + " / " + message11.Content);
            Console.WriteLine("message21: " + message21.Title + " / " + message21.Content);
            Console.WriteLine("message22: " + message22.Title + " / " + message22.Content);
        }
    }

    public sealed class MessageDetails : IAmImmutable
    {
        public MessageDetails(string title, string content)
        {
            this.CtorSet(_ => _.Title, title);
            this.CtorSet(_ => _.Content, content);
        }

        public string Title { get; }
        public string Content { get; }
    }
}
```

When I open the index.html file (that Bridge builds into bin/Debug/bridge by default), then I see the following in the browser console:

> Hi!
> message11: Title1 / Content1
> message21: Title2 / Content1
> message22: Title2 / Content2

If the "With" call was only setting one property and then leaving the others unset then  would have expected the "message21" line to read "Title2 / " as it would have had a blank Content value. But that is not the case.

Can you try to put together a small reproduce case to illustrate the problem? Failing that, if you can upload your project so far somewhere then I could have a look but that will take longer than checking a small sample!

**Comment by dnaustin** [comment]
I cannot seem to reproduce the error so I have uploaded my project [here]( https://github.com/dnaustin/Bridge-React-Demo)

**Comment by ProductiveRage** [comment]
I tried this code sample by adding the following packages to my project:

- ProductiveRage.Immutable.Extensions (this brings in Bridge.React)
- Bridge.ReactLoader (because this will pull in the React library without me needing to add script tags)

.. and I changed the line:

`var container = Document.GetElementById("main");`

to:

```
var container = Document.CreateElement("div");
Document.Body.AppendChild(container);
```

And the output that I got when I ran it was:

> Hi!
> message:  / 
> message1: Title1 / 
> message21: Title2 / 
> message22: Title2 / Content2

.. and that looks correct to me. What output did you get?

(I haven't tried your full code sample yet but I will if we don't get anywhere with a smaller reproduce case like this)

**Comment by ProductiveRage** [comment]
Right, I think I've found the problem! When I opened your example solution, I noticed that you're using the default bridge.json configuration that was introduced in Bridge 16. This includes the following line:

`"autoProperty": "Plain",`

For some reason, that breaks how the Immutable library works. You can work around it by either removing that line entirely from your file or by changing it to:

`"autoProperty": "Managed",`

In the [second post in the series,](https://www.productiverage.com/writing-react-apps-using-bridgenet-the-dan-way-from-first-principles) I recommend setting your bridge.json file to just this:

```
{
  "output": "Bridge/output",
  "combineScripts":  true
}
```

It looks like you skipped that step! :P

It does reveal a problem, though - the Immutable library *should* work with that "autoProperty" option (particularly since the Bridge Team set it to "plain" in the default file that they include in new Bridge projects). I think that I had it working in Bridge 16 but perhaps it's broken again in Bridge 17.

I will try to get a proper fix into the library but it looks like you can work around the issue for now by changing your bridge.json file.

**Comment by dnaustin** [comment]
Yes that fixed it! I must have misread that bit and just amended my file instead of completely replacing it. Thanks for your help :)
